certbot/filelock.py

```diff
@@ -0,0 +1,334 @@
+"""
+Library to lock a directory or a file using a lockfile.
+This implementation heavily relies on the work of benediktschmitt
+(https://github.com/benediktschmitt/py-filelock)
+and should be considered as a fork of his project.
+"""
+import atexit
+import logging
+import os
+import threading
+
+try:
+    import msvcrt
+except ImportError:
+    msvcrt = None  # type: ignore
+
+try:
+    import fcntl
+except ImportError:
+    fcntl = None  # type: ignore
+
+from certbot import errors
+from acme.magic_typing import List  # pylint: disable=unused-import, no-name-in-module
+
+logger = logging.getLogger(__name__)
+
+# Handling exit
+# ~~~~~~~~~~~~~
+_INITIAL_PID = os.getpid()
+_LOCKS = []  # type: List[BaseFileLock]
+
+
+def _release_all_locks():
+    """Release all locks acquired by FileLock."""
+    if _INITIAL_PID == os.getpid():
+        for lock in _LOCKS:
+            if lock.is_locked:
+                try:
+                    lock.release()
+                    logger.debug('Lock released: %s', lock.lock_file)
+                except (OSError, IOError):  # pragma: no cover
+                    logger.error('Exception occurred releasing lock: %s',
+                                 lock.lock_file, exc_info=True)
+
+
+# Every lock is released at least when Certbot exit.
+atexit.register(_release_all_locks)
```

Comment:
Yes, the duplication is here because of cyclic dependencies between util and filelock. I think it is acceptable however, it is only about 3 lines to import atexit, get PID and test PID in method.